imageJPG.ml

```diff
@@ -17,71 +17,308 @@
+
+        let ac_table_index = table_indexes land 0x0F in
+
+        if ac_table_index > 1 then
+          raise (Not_yet_implemented "Baseline decoding requires only two AC tables");
+
+        if !debug then
+          Printf.printf "Component Spec: component %d, DC table %d, AC table %d\n%!" component_index dc_table_index ac_table_index;
+        let component_spec = {
+          component_index;
+          dc_table_index;
+          ac_table_index;
+        } in
+        parse_component_spec ich (component_spec :: acc) (count - 1)
+    in
+    let length = get_bytes ich 2 |> int_of_str2_be in
+    let compspeccount = get_bytes ich 1 |> int_of_str1 in
+
+    if length <> (6 + 2*compspeccount) then
+      raise (Corrupted_image "Length does not match component specifier count");
+
+    let component_specs = parse_component_spec ich [] compspeccount in
+    let first_dct_component = get_bytes ich 1 |> int_of_str1 in
+
+    if first_dct_component <> 0 then
+      raise (Not_yet_implemented "Baseline decoding requires the first DCT component to be zero.");
+
+    let last_dct_component = get_bytes ich 1 |> int_of_str1 in
+
+    if last_dct_component <> 63 then
+      raise (Not_yet_implemented "Baseline decoding requires the last DCT component to be 63");
+
+    let approx = get_bytes ich 1 |> int_of_str1 in
+    let approx_high = (approx land 0xF0) lsr 4 in
+
+    if approx_high <> 0 then
+      raise (Not_yet_implemented "Baseline decoding requires the previous point transform to be zero");
+
+    let approx_low = approx land 0x0F in
+
+    if approx_low <> 0 then
+      raise (Not_yet_implemented "Baseline decoding requires the current point transform to be zero");
+
+    if !debug then
+      Printf.printf "SOS: length %d, component spec count: %d, first DCT component: %d, last DCT component: %d, approx high: %d, approx low: %d\n%!"
+        length compspeccount first_dct_component last_dct_component approx_high approx_low;
+    { component_specs; first_dct_component; last_dct_component; approx_high; approx_low }
+
+  (* Based on https://stackoverflow.com/a/48488655 *)
   let size ich =
-    let headers = read_header_data ich in
-    close_chunk_reader ich;
-
-    let rec find_SOF ls =
-      match ls with
-      | []        -> raise (Corrupted_image "No SOFn marker...");
-      | (m,d)::bs -> if m >= 0xc0 && m <= 0xcf && m <> 0xc4 && m <> 0xcc
-                     then (m - 0xc0, d)
-                     else find_SOF bs
+    let rec handle_marker ich c =
+      match string_of_marker c with
+      | "EOI" -> raise Not_found
+      | "SOF0" | "SOF1"  | "SOF2"  | "SOF3"  | "SOF5"  | "SOF6"  | "SOF7"
+      | "SOF9" | "SOF10" | "SOF11" | "SOF13" | "SOF14" | "SOF15" ->
+        let sof = parse_sof ich in
+        (sof.width, sof.height)
+      | marker ->
+        (if not (marker_stands_alone marker) then
+           let len = get_bytes ich 2 |> int_of_str2_be in
+           let data = get_bytes ich (len - 2) in
```

Comment:
maybe a guard against `len < 2` here?

Reply:
Technically actually a guard against `len < 4`, because JPEG lengths include themselves, but at that point `get_bytes` would raise `End_of_file` which gets caught at the decoder entry.

Reply:
Intuitively I think that *here* we have a chance of throwing `Corrupted_image`, where `get_bytes` doesn't really know what we're doing and can only throw a `Invalid_argument` or whatever, but I don't have a strong opinion on this, and am fine with making `get_bytes` responsible for checking input.